aws_ec2 inventory: add includes_filters and excludes_filters

goneri opened the pull request:
Expose a new `include_filters` and `exclude_filters` configuration keys that gives the user the ability to compose the final inventory list with several queries.

Closes: https://github.com/ansible-collections/amazon.aws/issues/248
Closes: https://github.com/ansible-collections/amazon.aws/issues/182

goneri commented:
Feedback on IRC (Cc: @flowerysong @tadeboro).

```
[15:06] <goneri> Feedback on https://github.com/ansible-collections/amazon.aws/pull/328 welcome.
[15:37] <flowerysong> goneri: I still think the correct interface for this is for filters: to accept a list, as discussed in https://github.com/ansible/ansible/pull/58132
[15:38] <flowerysong> goneri: Something like https://github.com/flowerysong/ansible-amazon.aws/commit/6546a299
[15:41] <tadeboro> I must agree with flowerysong here. Introducting yet another option that just adds hosts to existing filter result feels awkward.
[16:25] <goneri> I would also add an excludes_entries_matching for this one https://github.com/ansible-collections/amazon.aws/issues/182
[16:26] <goneri> in this case the two options would make sense. I don't really like when the behaviour of an option changes because of the type.
[16:26] <goneri> This leads to confusion.
```

flowerysong commented:
Having multiple options that do the same thing leads to even more confusion. If you really think that it's confusing, you can deprecate the old way of passing filters and require that the user specify a list.

goneri commented:
```
[16:30] <tadeboro> For ServiceNow, we implemented https://github.com/ansible-collections/servicenow.itsm/blob/main/docs/servicenow.itsm.servicenow.itsm.now_inventory.rst#examples
[16:30] <goneri> I updated https://github.com/ansible-collections/amazon.aws/pull/328 with your comments
[16:30] <goneri> I guess it's better to continue there.
[16:30] <flowerysong> It doesn't change the behaviour, it just automatically wraps the dict in a list (similar to other options that accept lists.)
[16:30] <tadeboro> We allow users to specify the inclusion and exclusion criteria.
[16:31] <goneri> flowerysong: in one case we've got a OR and another one a AND.
[16:33] <goneri> tadeboro: interesting. I see the includes/excludes keys as a similarity, am I right?.
[16:34] <tadeboro> If you want to also introduce an exclusion criteria, then deprecating the filter and replacing it with something else maybe makes the most sense? Something like `include_filters` and `exclude_filters` maybe?
[16:35] <tadeboro> In ServiceNow's case, we structured things a bit dfferently because we combine inclusions and exclusion into a single query and then just partition the returned hosts.
[16:35] <goneri> Yes! I agree, includes_entries_matching is pretty lame :-)
[16:36] <goneri> Yes, I understand. With AWS, we need to do one request per filter, it's a bit more complicated.
[16:37] <goneri> I suggest to refresh my PR with your name suggestion: include_filters and exclude_filters. Would this work for you?
[16:38] <tadeboro> It would be a bit strange if I sad no after I suggested the names ;)
[16:38] <goneri> Great, I will push a PR tomorrow. Thanks flowerysong and tadeboro.
```

goneri commented:
> Having multiple options that do the same thing leads to even more confusion. If you really think that it's confusing, you can deprecate the old way of passing filters and require that the user specify a list.

I don't think we need to deprecate the current `filters` and for the users to adjust their configurations. These two news `include_filters` and `exclude_filters` would only accept list.

tremble commented on plugins/inventory/aws_ec2.py:
```diff
@@ -51,6 +51,14 @@
           default: {}
+        includes_entries_matching:
+          description:
+              - A list of filters. The instances matching at least one of the filters are included in the result.
```
I needed to read this twice to parse it properly.  Oddly enough I found the changelog easier to understand than this description.

```suggestion
              - A list of filters. Any instances matching at least one of the filters are included in the result.
```

goneri replied:
:-), thank you for the feedback.

tremble commented on plugins/inventory/aws_ec2.py:
```diff
@@ -51,6 +51,14 @@
+          description:
+              - A list of filters. The instances matching at least one of the filters are included in the result.
+              - Available filters are listed here U(http://docs.aws.amazon.com/cli/latest/reference/ec2/describe-instances.html#options).
+              - Every entry of this list trigger a search query. This is the reason why from a performance point of view, it's better to
```
```suggestion
              - Every entry in this list triggers a search query.  As such, from a performance point of view, it's better to
```

tremble commented on plugins/inventory/aws_ec2.py:
```diff
@@ -456,7 +478,7 @@ def _get_instances_by_region(self, regions, filters, strict_permissions):
             all_instances.extend(instances)
 
-        return sorted(all_instances, key=lambda x: x['InstanceId'])
+        return all_instances
```
While Instance ID isn't necessarily the right thing to sort by, dropping the sort may cause problems for anyone who relied on this returning a predictable inventory.  By default Ansible will honour the order that hosts are listed in the inventory.

goneri replied:
I would like to preserve the order. I just moved the line in `_query()`.

tremble replied:
Ah, ok I missed that :)

tremble commented on tests/integration/targets/inventory_aws_ec2/runme.sh:
```diff
@@ -38,6 +38,9 @@ ansible-playbook playbooks/create_inventory_config.yml -e "template='inventory_w
 ansible-playbook playbooks/test_populating_inventory_with_constructed.yml "$@"
 ansible-playbook playbooks/create_inventory_config.yml -e "template='inventory_with_concatenation.yml.j2'" "$@"
 ansible-playbook playbooks/test_populating_inventory_with_concatenation.yml "$@"
+# generate inventory config with includes_entries_matching and prepare the tests
+ansible-playbook playbooks/create_inventory_config.yml -e "template='inventory_with_includes_entries_matching.yml.j2'" "$@"
+ansible-playbook playbooks/test_populating_inventory_with_includes_entries_matching.yml "$@"
```
It looks like the template and playbook are missing.

I'm not sure why, but it looks like the integration tests didn't fire properly

goneri replied:
:thinking: indeed.

tremble commented on plugins/inventory/aws_ec2.py:
```diff
@@ -51,6 +51,22 @@
+          default: []
+        exclude_filters:
+          description:
+              - A list of filters. Any instances matching one of the filters are excluded from the result.
```
Needs to explain if include or exclude takes priority.  From the code it looks like exclude takes priority over include.

goneri replied:
Yes,g ood point.

IPvSean replied:
agreed, if i put like 

```
hostnames:
- tag:Name
  - my_host 
```

but that instance `my_host` was also under exclude_filters what would happen?

goneri replied:
Exclude prevails, you would get an empty list.

tremble commented on plugins/inventory/aws_ec2.py:
```diff
@@ -537,14 +567,35 @@ def _get_hostname(self, instance, hostnames):
-            :param filters: a list of boto3 filter dictionaries
+            :param include_filters: a list of boto3 filter dictionaries
+            :param exclude_filters: a list of boto3 filter dictionaries
             :param hostnames: a list of hostname destination variables in order of preference
+            :param includes_entries_matching: a list of filters, instances matching these filters are included in the result
```
```suggestion
```

hostnames and includes_entries_matching look to be gone?

goneri replied:
`include_filters` and `exclude_filters` supersede` includes_entries_matching`. For `hostnames` it's a mistake.

tremble commented on plugins/inventory/aws_ec2.py:
```diff
@@ -537,14 +567,35 @@ def _get_hostname(self, instance, hostnames):
+                    ansible_dict_to_boto3_filter_list(filter),
+                    strict_permissions):
+                ids_to_ignore.append(i['InstanceId'])
+        for filter in include_filters:
+            for i in self._get_instances_by_region(
+                    regions,
+                    ansible_dict_to_boto3_filter_list(filter),
+                    strict_permissions):
+                if i['InstanceId'] not in ids_to_ignore:
+                    instances.append(i)
```
While Ansible deals well with duplicates in an inventory, it's probably better to try and filter them out.  Maybe insert into a map and then pull them back out into a list while sorting?

goneri replied:
Addressed by 0869bf162e9f3f37a1afb5f2a781e7ccd1a90f6a

tremble commented on plugins/inventory/aws_ec2.py:
```diff
@@ -649,7 +700,8 @@ def parse(self, inventory, loader, path, cache=True):
         # get user specifications
         regions = self.get_option('regions')
-        filters = ansible_dict_to_boto3_filter_list(self.get_option('filters'))
+        include_filters = [self.get_option('filters')] + self.get_option('include_filters')
+        exclude_filters = self.get_option('exclude_filters')
```
Should we deprecate the 'filters' option ?  (can we deprecate inventory parameters?)

goneri replied:
I don't feel the need for that. The long term support should be rather easy anyway.

tremble commented on tests/integration/targets/inventory_aws_ec2/playbooks/test_populating_inventory_with_include_or_exclude_filters.yml:
```diff
@@ -0,0 +1,118 @@
+
+        # Create new host, refresh inventory
+        - name: create a new host (1/3)
+          ec2:
```
The ec2 module is boto v2 based and handles rate-limiting poorly (resulting in test flakes).

I believe ec2_instance is on the priority list to migrate to amazon.aws I'd recommend using it instead.

goneri replied:
I just follow the de-facto convention for this test role. I would prefer to change that in another PR and take care of all the `ec2_instance` calls.

tremble replied:
FYI I've already made that change in community.aws

tremble commented on tests/integration/targets/inventory_aws_ec2/templates/inventory_with_include_or_exclude_filters.yml.j2:
```diff
@@ -0,0 +1,21 @@
+  - '{{ resource_prefix }}_3'
+include_filters:
+- tag:Name:
+  - '{{ resource_prefix }}_2'
+exclude_filters:
+- tag:Name:
+  - '{{ resource_prefix }}_3'
```
While the logic paths are currently the same, I'd recommend adding an _4 host too.  To handle someone messing with the logic in future.

```suggestion
  - '{{ resource_prefix }}_2'
  - '{{ resource_prefix }}_4'
exclude_filters:
- tag:Name:
  - '{{ resource_prefix }}_3'
  - '{{ resource_prefix }}_4'
```

tremble commented:
Mostly looks good, primarily nit-picking at this point.

jillr commented:
Thanks very much @goneri, this looks good.  Can you please work with the Tower team to schedule this with QE?
